File: fb/128.longest-consecutive-sequence.cpp
// Time O(n)
// Space O(n)
class Solution {
public:
    int longestConsecutive(vector<int>& nums) {
        unordered_set<int> numsSet(nums.begin(), nums.end());
        int longestStreak = 0;
        for(auto i : numsSet) {
            // only start searching when i is the begining (i.e. smallest)
            // of a sequence to avoid redundent searching
            if(!numsSet.count(i - 1)) {
                int currentStreak = 1;
                int curr = i + 1;
                while(numsSet.count(curr)) {
                    currentStreak++;
                    curr++;
                }
                longestStreak = max(currentStreak, longestStreak);
            }
        }
        return longestStreak;
    }
};

File: fb/128.longest-consecutive-sequence.py
class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        longestStreak = 0
        numSet = set(nums)
        
        for num in nums:
            if num-1 not in numSet:
                localStreak = 1
                curNum = num
                
                while curNum+1 in numSet:
                    localStreak += 1
                    curNum += 1
                    
                longestStreak = max(longestStreak, localStreak)
        return longestStreak
